module/zfs/dmu_recv.c

```diff
@@ -1236,6 +1289,185 @@ dmu_recv_begin(char *tofs, char *tosnap, dmu_replay_record_t *drr_begin,
+			dsl_dataset_t *ds;
+			dsl_crypto_key_t *dck = NULL;
+			uint8_t salt[ZIO_DATA_SALT_LEN];
+			uint8_t iv[ZIO_DATA_IV_LEN];
+			uint8_t mac[ZIO_DATA_MAC_LEN];
+			boolean_t no_crypt = B_FALSE;
+			dsl_pool_t *dp = dmu_objset_pool(rwa->os);
+			abd_t *eabd = abd_alloc_linear(drrw->drr_logical_size,
+			    B_FALSE);
+			zio_crypt_decode_params_bp(bp, salt, iv);
+			zio_crypt_decode_mac_bp(bp, mac);
+
+			dsl_pool_config_enter(dp, FTAG);
+			err = dsl_dataset_hold_flags(dp, rwa->tofs,
+			    DS_HOLD_FLAG_DECRYPT, FTAG, &ds);
+			if (err != 0) {
+				dsl_pool_config_exit(dp, FTAG);
+				abd_free(eabd);
+				return (SET_ERROR(EACCES));
+			}
+
+			/* Look up the key from the spa's keystore */
+			err = spa_keystore_lookup_key(rwa->os->os_spa,
+			    zb.zb_objset, FTAG, &dck);
+			if (err != 0) {
+				dsl_dataset_rele_flags(ds, DS_HOLD_FLAG_DECRYPT,
+				    FTAG);
+				dsl_pool_config_exit(dp, FTAG);
+				abd_free(eabd);
+				return (SET_ERROR(EACCES));
+			}
+
+			err = zio_do_crypt_abd(B_TRUE, &dck->dck_key,
+			    BP_GET_TYPE(bp), BP_SHOULD_BYTESWAP(bp), salt, iv,
+			    mac, drrw->drr_logical_size, abd, eabd, &no_crypt);
+
+			spa_keystore_dsl_key_rele(rwa->os->os_spa, dck, FTAG);
+			dsl_dataset_rele_flags(ds, DS_HOLD_FLAG_DECRYPT, FTAG);
+			dsl_pool_config_exit(dp, FTAG);
+
+			ASSERT0(no_crypt);
+			if (err != 0) {
+				abd_free(eabd);
+				return (err);
+			}
+			/* Swap in the newly encrypted data into the abd */
+			abd_free(abd);
+			abd = eabd;
+
+			/*
+			 * We want to prevent zio_rewrite() from trying to
+			 * encrypt the data again
+			 */
+			flags |= ZIO_FLAG_RAW;
+		}
+
+		if (BP_GET_COMPRESS(bp) != ZIO_COMPRESS_OFF) {
+			/* Recompress the data */
+			abd_t *cabd = abd_alloc_linear(drrw->drr_logical_size,
+			    B_FALSE);
+			zio_compress_data(BP_GET_COMPRESS(bp),
+			    abd, abd_to_buf(cabd), drrw->drr_logical_size,
+			    rwa->os->os_complevel);
+			/* Swap in the newly compressed data into the abd */
+			abd_free(abd);
+			abd = cabd;
+		}
+
+		rrd->abd = abd;
+	}
+
+	io = zio_rewrite(NULL, rwa->os->os_spa, 0, bp, abd, BP_GET_PSIZE(bp),
+	    NULL, NULL, ZIO_PRIORITY_SYNC_WRITE, flags, &zb);
+
+	/* compute new bp checksum value and make sure it matches the old one */
+	zio_checksum_compute(io, BP_GET_CHECKSUM(bp), abd,
+	    drrw->drr_logical_size);
+	if (drrw->drr_logical_size != BP_GET_LSIZE(bp) ||
+	    !ZIO_CHECKSUM_EQUAL(bp_cksum, io->io_bp->blk_cksum)) {
+		zfs_dbgmsg("corrected block checksum is wrong or %lld != %lld",
+		    drrw->drr_logical_size, BP_GET_PSIZE(bp));
+		zio_destroy(io);
+
+		if (zfs_recv_best_effort_corrective != 0)
+			return (0);
+		return (SET_ERROR(ECKSUM));
+	}
+
+	/* Correct the corruption in place */
+	err = zio_wait(io);
+	if (err == 0) {
+		cr_cb_data_t *cb_data =
+		    kmem_alloc(sizeof (cr_cb_data_t), KM_SLEEP);
+		cb_data->spa = rwa->os->os_spa;
+		cb_data->size = drrw->drr_logical_size;
+		cb_data->zb = zb;
+		/* Test if healing worked by re-reading the bp */
```

Comment:
If this re-read generates ECKSUM because the healing didn't work, it just stays as a checksum error?  Maybe the zfstest can cover this?

Reply:
It does stay a checksum error yeah. I'm not sure how I could produce a healing failure for zfs-test, any ideas on that?

Reply:
The way it's written now this function does return an error if the zio_read() fails and zfs_recv_best_effort_corrective is not set to zero (which is the default). For testing, I couldn't come up with a way to selectively corrupt a healed block but I did test this path by just forcing the functions to return ECKSUM as you suggested and the results were reasonable:
The corrective recv returned `cannot receive new filesystem stream: corrective receive was not able to reconstruct the data needed for healing.` msg and I was able to unload the module ok (without leak etc)

Reply:
I'm only seeing "err" set from the zio_rewrite.  The zio_read is with zio_nowait, so I don't see it setting/returning err.  It's good that corrective_read_done preserves the spa error if the zio_read fails though.

Reply:
oh i see what you mean now. I wasn't sure if it was better to use async re-read there but since we just wrote that block it probably does make sense to wait for the re-read and return it's results. I've updated the code to do that.
The only downside I can think of is if there is a lot of corruption the healing recv will run slower if we're not using async re-reads.